### ISSUE TITLE
Serverless ACA-Py Agents Proposal

### ISSUE THREAD

**Issue by rtatton** [hyperledger/aries-cloudagent-python#1648]
# Context
[Serverless computing](https://en.wikipedia.org/wiki/Serverless_computing) can provide cost-savings and scalability for many applications. One such offering in this paradigm is function-as-a-service (FaaS), which allows developers to focus on application logic, without needing to worry about server maintenance, scaling, load balancing, or cost inefficiencies. Of course, [no lunch is free](https://en.wikipedia.org/wiki/There_ain't_no_such_thing_as_a_free_lunch); these functions have limited compute capabilities and can typically only run for relatively short durations (i.e., 15 minutes). In spite of these drawbacks, deploying ACA-Py agents in a serverless environment could save development time and money, which makes it an interesting possibility.

Because serverless agents can time out, the current (as of writing this issue) developments of persistent queues using Redis and Kafka (see issue #1513 and references therein) are quite relevant and compatible with this design. In fact, the issue that persistent queues are meant to address, as stated in #1513, can more broadly be solved with serverless approaches. As stated below, another queue implementation using Amazon SQS is proposed because of its advantages over the currently proposed solutions.

# Architecture
Partly a result of my own development experience and bias, the following design uses Amazon Web Services (AWS) Lambda Function and Amazon Simple Service Queue (SQS); though, it is likely also possible to implement a similar design using [Microsoft Azure Functions](https://docs.microsoft.com/en-us/azure/azure-functions/functions-overview) or [Google Cloud Functions](https://cloud.google.com/functions/) (an exercise left to the reader, I suppose).

![Diagram - Serverless Aries Agent](https://user-images.githubusercontent.com/38771770/156301999-fa3a3c01-eda1-4b14-ad55-2f95d4fd1046.png)

**Comment by swcurran** [comment]
This is pretty cool @rtatton -- nice work.  Would you be interested in presenting this at an ACA-Pug meeting?  The next meeting (next week) probably has too much content already, but perhaps the next one (3 weeks)?

**Comment by tankcdr** [comment]
Agreed. This is really cool and inline with what I want my team to accomplish...but with Azure :) Have you thought of using Step Functions? I was thinking along the lines of using Durable Functions (Azure equivalent, kind of) to fetch config first and pass to next function to start agent and process messages.

**Comment by rtatton** [comment]
@tankcdr That's awesome! I had not thought about using Step Functions to invoke the Lambda functions. Right now, anything with the appropriate access control can send a message to an agent's queue.

While conceptually straightforward, the idea of a one-to-one mapping between agent and Lambda function means that there is a limited number of Lambda functions that can be spun up in a given account. Being able to somehow ensure that account-level concurrency is not violated as you continue to add more agents to a given AWS account is still an aspect of this approach that needs to be resolved.

For reference, [here](https://github.com/project-cirrus/infrastructure) is the repository I am working in to implement this functionality. I am using AWS CDK so that anyone with an AWS account can run a few CLI commands to get an instance of the service up and running. An API is exposed which allows authorized users to upload controller code, create nodes and, delete nodes, where a node is defined in the above diagram,
```
node = (Lambda function with framework and controller) + (SQS queue)
```

**Comment by swcurran** [comment]
@rtatton is going to talk about this at the next ACA-Pug meeting on April 5th, 8AM Pacific/5PM CET.

https://wiki.hyperledger.org/display/ARIES/2022-04-05+Aries+Cloud+Agent+-+Python+Users+Group+Community+Meeting

@tankcdr -- great if you could join as well and cover what you are doing with Azure as well.

Thanks!

**Comment by ryebread321** [comment]
For context, I'm attaching the project report that was written.

[project-report.pdf](https://github.com/user-attachments/files/24595730/project-report.pdf)

Since there is no intent on completing this implementation, I'm closing this issue.

**Comment by swcurran** [comment]
@ryjones -- seems possible (likely?) that the link above to the "project-report.pdf" is a phishing attack.  Note the change in the submitter GitHub name -- used to be @rtatton.  I definitely would not open the PDF.  Do you have a way of reporting it?

**Comment by ryebread321** [comment]
> [@ryjones](https://github.com/ryjones) -- seems possible (likely?) that the link above to the "project-report.pdf" is a phishing attack. Note the change in the submitter GitHub name -- used to be @rtatton. I definitely would not open the PDF. Do you have a way of reporting it?

My apologies for the concern. I changed my GitHub username last year. I uploaded the PDF directly via GitHub’s file attachment feature, which you can verify based on the URL.
